Modules/pyexpat.c

```diff
@@ -1654,7 +1654,18 @@ static int
 add_error(PyObject *errors_module, PyObject *codes_dict,
           PyObject *rev_codes_dict, const char *name, int value)
 {
-    const char *error_string = XML_ErrorString(value);
+    const char * const error_string_or_null = XML_ErrorString(value);
+
+    /* Be robust towards runtime libexpat being too old to know
+     * all the latest error codes */
+    const char *error_string = error_string_or_null;
+    char error_string_substitute[100];
+    if (error_string_or_null == NULL) {
```

Comment:
We can use to the previous code right since version guard prevents the situation?

`````

const char *error_string = XML_ErrorString(value);
`````

Reply:
This is where the difference between runtime and build time comes into play, as explained at https://github.com/python/cpython/pull/30188#discussion_r775037301 . I'm good either way, I just need a decision if you want robust or simple.

Reply:
I prefer the following style but I am not an expert on distro's issues.
So I would like to request the code review for this part to @tiran and @vstinner :)

`````c
const char *error_string = XML_ErrorString(value);
if (error_string == NULL) {
     return -1;
}
if (PyModule_AddStringConstant(errors_module, name, error_string) < 0) {
    return -1;
}
`````

Reply:
> I prefer the following style [..]

This will make the module import fail. That seems suboptimal. What's the idea, why the preference?
For proof:
```console
# ./python -c 'import xml.parsers.expat.errors'
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "[..]/cpython/Lib/xml/parsers/expat.py", line 4, in <module>
    from pyexpat import *
    ^^^^^^^^^^^^^^^^^^^^^
SystemError: execution of module pyexpat failed without setting an exception
```

Reply:
@hartwork 
Import fail might be if build environment != runtime environment. right?
It will fail if the environment is different, I just want to know the CPython of the distro version executed with the different runtime version.
If the normal distro version guarantee the build environment and runtime environment is the same until the user does not tweak libexpat version, it will be okay.
IMHO, if users want to run different libexpat version, they should rebuild CPython, that looks right away.

Early fail which I suggested is a pretty usual workaround in CPython module initialization.

`````

@@ -1658,12 +1658,9 @@ add_error(PyObject *errors_module, PyObject *codes_dict,

     /* Be robust towards runtime libexpat being too old to know
      * all the latest error codes */
-    const char *error_string = error_string_or_null;
-    char error_string_substitute[100];
-    if (error_string_or_null == NULL) {
-        sprintf(error_string_substitute,
-                "error code %i too recent for runtime libexpat", value);
-        error_string = error_string_substitute;
+    const char *error_string = XML_ErrorString(value);
+    if (error_string == NULL) {
+        return -1;
     }

     if (PyModule_AddStringConstant(errors_module, name, error_string) < 0) {

$> cpython git:(pr/30188) ✗ ./python.exe -c 'import xml.parsers.expat.errors'
`````

Reply:
So if distro version is built with vendor version, and run with system installed library version,
my suggestion will be wrong.

Reply:
@corona10 I think the most likely case is building CPython with `--with-system-expat` and then using the resulting CPython binaries with older libexpat binaries at runtime, e.g. in context of a binary Linux distro with multiple release branches. If everyone agrees that that scenario is meant to be 100% unsupported and okay break to pyexpat, then we can just go with `return -1` on a `NULL` error string.

Reply:
@hartwork 
I check the Linux distro which uses libexpat 1.6.7 it works as expected (module import success + use errr code for specific version)

I would like to suggest to changing it to return -1.

Reply:
> @hartwork I check the Linux distro which uses libexpat 1.6.7 it works as expected (module import success + use errr code for specific version)

1.6.7 is a is a shared library file version, not a version of libexpat, and not trivial to "decode". Let's stick to release and packaging versions please, so that everyone knows which versions are being referred to.


> I would like to suggest to changing it to return -1.

Please elaborate why. Thanks.

Reply:
@hartwork 

My opinion is that supporting error codes should not be added.
https://github.com/python/cpython/blob/196b53eb1e62871ca80dee180e4891b4dd5c52ac/Modules/_sqlite/module.c#L303
For example, sqlite3 uses a similar approach, the only difference is they do not extract value at runtime.
If you think that the error_string should be filled even if the runtime environment is different. Please go ahead.
But does it guarantee normal behavior? if so I am okay with filling fallback string. 
Even if we choose any options I will leave this PR to other core dev to listen to their opinion also.
Since this patch will affect a lot of distros.

Reply:
Hi @corona10 ,

I have dropped the `NULL` handling for now, because there are multiple places where `XML_ErrorString` is called and if someone decided that a runtime-build-time-mismatch is a problem, all of these places would need `NULL` handling:

```console
# git --no-pager grep 'XML_ErrorString(' -- Modules/pyexpat.c
Modules/pyexpat.c:                                  XML_ErrorString(code), lineno, column);
Modules/pyexpat.c:    return Py_BuildValue("z", XML_ErrorString((int)code));
Modules/pyexpat.c:    const char *error_string = XML_ErrorString(value);
```

> https://github.com/python/cpython/blob/196b53eb1e62871ca80dee180e4891b4dd5c52ac/Modules/_sqlite/module.c#L303
> 
> For example, sqlite3 uses a similar approach, the only difference is they do not extract value at runtime.

That runtime part difference is key, though. With only compile time, there cannot be a build/run mismatch.

> If you think that the error_string should be filled even if the runtime environment is different. Please go ahead.
> But does it guarantee normal behavior? if so I am okay with filling fallback string.

Interesting point, I guess that's a question of view.